### ISSUE TITLE
training time

### ISSUE THREAD

**Issue by lixiang-ucas** [GeorgeSeif/FC-DenseNet-Tiramisu#1]
How long does it take to train this model? 
When i run this code, i got the following messages and the training process is extremely slow, why?

The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-12-30 02:23:19.780078: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
This model has 9270812 trainable parameters
***** Begin training *****
[2017-12-30 02:31:59] Epoch = 0 Count = 20 Current = 2.52 Time = 25.51
[2017-12-30 02:40:30] Epoch = 0 Count = 40 Current = 2.36 Time = 25.65
[2017-12-30 02:49:02] Epoch = 0 Count = 60 Current = 2.20 Time = 25.60
[2017-12-30 02:57:29] Epoch = 0 Count = 80 Current = 2.02 Time = 25.34
[2017-12-30 03:05:56] Epoch = 0 Count = 100 Current = 1.81 Time = 25.45
[2017-12-30 03:14:23] Epoch = 0 Count = 120 Current = 1.42 Time = 25.31
[2017-12-30 03:22:50] Epoch = 0 Count = 140 Current = 1.61 Time = 25.34
[2017-12-30 03:31:18] Epoch = 0 Count = 160 Current = 1.49 Time = 25.29
[2017-12-30 03:39:44] Epoch = 0 Count = 180 Current = 1.65 Time = 25.30
[2017-12-30 03:48:13] Epoch = 0 Count = 200 Current = 1.42 Time = 25.48
[2017-12-30 03:56:43] Epoch = 0 Count = 220 Current = 1.50 Time = 25.55
[2017-12-30 04:05:14] Epoch = 0 Count = 240 Current = 1.60 Time = 25.81
[2017-12-30 04:13:48] Epoch = 0 Count = 260 Current = 1.46 Time = 25.56
[2017-12-30 04:22:18] Epoch = 0 Count = 280 Current = 1.89 Time = 25.88
[2017-12-30 04:30:49] Epoch = 0 Count = 300 Current = 1.46 Time = 25.55
[2017-12-30 04:39:22] Epoch = 0 Count = 320 Current = 1.32 Time = 25.33

**Comment by GeorgeSeif** [comment]
Hi there,

Which GPU are you using? For your reference, I'm running on a 1080 Ti and have the following training output:

[2017-12-29 16:44:06] Epoch = 0 Count = 20 Current = 2.21 Time = 0.45
[2017-12-29 16:44:15] Epoch = 0 Count = 40 Current = 2.03 Time = 0.45
[2017-12-29 16:44:24] Epoch = 0 Count = 60 Current = 2.01 Time = 0.45
[2017-12-29 16:44:33] Epoch = 0 Count = 80 Current = 1.92 Time = 0.45
[2017-12-29 16:44:42] Epoch = 0 Count = 100 Current = 1.98 Time = 0.45
[2017-12-29 16:44:51] Epoch = 0 Count = 120 Current = 2.15 Time = 0.45
[2017-12-29 16:45:00] Epoch = 0 Count = 140 Current = 1.79 Time = 0.45
[2017-12-29 16:45:09] Epoch = 0 Count = 160 Current = 1.65 Time = 0.45
[2017-12-29 16:45:18] Epoch = 0 Count = 180 Current = 1.34 Time = 0.45

As for the output message, that just means you don't have TensorFlow compiled with those CPU speed-up instructions. I'm not sure how much of a difference it would make on a CPU **only** training, but on GPU there won't be much of a difference

**Comment by lixiang-ucas** [comment]
Thanks. I'm using a GTX TITAN X. I'll update my tensorflow and try again.

**Comment by GeorgeSeif** [comment]
You shouldn't have to with a Titan X. Our speed shouldn't be that different. Check out this benchmark:
https://github.com/jcjohnson/cnn-benchmarks

Perhaps yours is running on the CPU? Because from 25 seconds to 0.45 is a big difference

**Comment by lixiang-ucas** [comment]
i guess so, maybe there's something wrong with my tensorflow-gpu installation, i'm trying.

**Comment by lixiang-ucas** [comment]
I reinstall my tensorflow and it just works fine now. Thank you for your suggestions.

**Comment by GeorgeSeif** [comment]
No problemo!

**Comment by roestigraben** [comment]
I am using CPU only and have a tensorflow version with all the new instructions compiled. The speedup is not so great (approx 10-20%). Also the newest version of Tensorflow (1.6) actually compiles already some of the instructions.

My training times are also  way too slow and i am looking into building a box (Ubuntu 16.04).
Your GPU's are very high end and give around  50x improvement as I see and which is nice.
I probably will settle with a GTX1080.

Question: what key parts specs do you guys use for your machine
CPU i5 8400 vs ?
RAM 16G vs ?
Motherboard Z370 vs ?                 (single GPU 16lanes PCIe 3.0 or double at 8 lanes PICe 3.0)

Thanks for sharing

**Comment by lixiang-ucas** [comment]
@roestigraben you'd better give up cpu for deep learning training, too slow!!!
I'm using Tesla K80 GPU with CPU E7 now.

**Comment by roestigraben** [comment]
Thanks
but the prices are sky high for the K80. I saw in excess of 6000$ while the GTX's are still OK with around 600$. Are the K80'2 10x faster?

**Comment by lixiang-ucas** [comment]
@roestigraben considering the price, GTX 1080 or 1080Ti is a good choice. For speed comparison, check out this benchmark:
https://github.com/jcjohnson/cnn-benchmarks

**Comment by roestigraben** [comment]
Thanks a lot! very interesting benchmark

**Comment by roestigraben** [comment]
Hello,
So, my CPU based training went way too slow with around 171s per set of 20 iteration with all the rest being default values(model, batch_size, ......) and despite the latest Tensorflow version 1.7 (which BTW improved speed by 10-20% versus v1.6 and older)
This compares to George's 0.45s and even lixiangs's 25s  (I hope we have the same setup for ll the other values)

I went to paperspace coud provider and ran the same with a Quadro P4000 with 1792 Cuda cores, 8 GB DDR5 and a memory bandwidth of 243Gb7s.
It ran very nicely indeed with about but the end result of 5s is disappointing.
George's GTX1080TI has 11GB of memory, ? cores but should be similar to the Quadro P4000.
10X difference seems enormous.
And, it runs on the GPU, not on the CPU 

Do you have any hints why this is?

**Comment by roestigraben** [comment]
BTW, I checked some of the tech specs a bit more
                              GTX1080Ti     Quadro P4000 has

DDR5 memory      11GB               8GB
Cuda Cores           3584              1792
performance         10.6                5.2TFLOPS
arch                       Pascal            Pascal       

Does this call for 10X difference?

**Comment by kkose** [comment]
Hi, 
I would like to ask what training speed (image/sec) should I expect to see with the toolkit. I am training with 256x256 images (batch size of 8) on a Titan X (12GB) GPU. I am using the default "FC-DenseNet56" model, which supposed to have ~1.5M parameters.  A sample from my training times are as follows

[2018-12-13 15:15:56] Epoch = 4 Count = 40 Current_Loss = 1.8203 Time = 3.99
[2018-12-13 15:16:00] Epoch = 4 Count = 80 Current_Loss = 1.5325 Time = 4.01
[2018-12-13 15:16:04] Epoch = 4 Count = 120 Current_Loss = 1.6215 Time = 3.94
...
This corresponds to 10 images/sec processing speed. 
I was able to train models with 6M parameters in Keras (with TF backend) at ~50 images/sec. I am not sure why native TF implementation is this slow. 

Any suggestions on which configuration parameters to look at in order to increase the speed of training. I reinstalled the tensorflow library (using 1.12, the current version) but that did not help. 
thanks
